credential/schema/vcjsonschema_test.go

```diff
@@ -49,11 +49,351 @@ func TestValidateCredentialAgainstSchema(t *testing.T) {
+			assert.NoError(t, err)
+		})
+
+		t.Run("$id is not a valid URI", func(t *testing.T) {
+			cred := getTestVCJSONSchemaCredential()
+			schema := getTestVCJSONSchemaSchema()
+
+			invalidURI := "bad"
+			schemaSubject := schema["credentialSubject"].(map[string]any)
+			schemaSubjectSchema := schemaSubject["jsonSchema"].(map[string]any)
+			schemaSubjectSchema["$id"] = invalidURI
+			schemaSubject["jsonSchema"] = schemaSubjectSchema
+			schema["credentialSubject"] = schemaSubject
+
+			err := IsCredentialValidForJSONSchema(cred, schema, JSONSchemaCredentialType)
+			assert.Error(t, err)
+			assert.Contains(t, err.Error(), "not a valid URI")
+		})
+	})
+
+	t.Run("the $schema property MUST be present", func(t *testing.T) {
+		t.Run("schema property is present and not empty", func(t *testing.T) {
+			cred := getTestVCJSONSchemaCredential()
+			schema := getTestVCJSONSchemaSchema()
+			schemaCredSubject := schema["credentialSubject"].(map[string]any)
+			schemaCredSubjectSchema := schemaCredSubject["jsonSchema"].(map[string]any)
+			assert.NotEmpty(t, schemaCredSubjectSchema["$schema"])
+			err := IsCredentialValidForJSONSchema(cred, schema, JSONSchemaCredentialType)
+			assert.NoError(t, err)
+		})
+
+		t.Run("schema property is not present", func(t *testing.T) {
+			cred := getTestVCJSONSchemaCredential()
+			schema := getTestVCJSONSchemaSchema()
+			schemaCredSubject := schema["credentialSubject"].(map[string]any)
+			schemaCredSubjectSchema := schemaCredSubject["jsonSchema"].(map[string]any)
+			delete(schemaCredSubjectSchema, "$schema")
+			schemaCredSubject["jsonSchema"] = schemaCredSubjectSchema
+			schema["credentialSubject"] = schemaCredSubject
+			err := IsCredentialValidForJSONSchema(cred, schema, JSONSchemaCredentialType)
+			assert.Error(t, err)
+			assert.Contains(t, err.Error(), "credential schema does not contain a `$schema` property")
+		})
+	})
+}
+
+func getTestJSONSchemaCredential() credential.VerifiableCredential {
+	return credential.VerifiableCredential{
+		ID:     "https://example.com/credentials/3734",
+		Issuer: " https://example.com/issuers/14",
+		CredentialSubject: map[string]any{
+			"id":           "did:example:ebfeb1f712ebc6f1c276e12ec21",
+			"emailAddress": "test@test.com",
+		},
+		CredentialSchema: &credential.CredentialSchema{
+			ID:   "https://example.com/schemas/email.json",
+			Type: "JsonSchema",
+		},
+	}
+}
+
+func getTestJSONSchemaSchema() VCJSONSchema {
+	return VCJSONSchema{
+		"$id":     "https://example.com/schemas/email.json",
+		"$schema": "https://json-schema.org/draft-07/schema#",
+		"title":   "Email address",
+		"type":    "object",
+		"properties": map[string]any{
+			"credentialSubject": map[string]any{
+				"type": "object",
+				"properties": map[string]any{
+					"emailAddress": map[string]any{
+						"type": "string",
+					},
+				},
+				"required": []string{"emailAddress"},
+			},
+		},
+		"required": []string{"credentialSubject"},
+	}
```

Comment:
I would suggest putting the `json` and unmarshalling, instead of instantiating a map.

Reply:
going to stick with a map since it's easier to maintain IMO